src/pallene/gc.lua

```diff
@@ -28,77 +29,234 @@ local types = require "pallene.types"
-    -- saved by the caller. Also note that we only care about variables, not about upvalues.
-    -- The latter are already exposed to the GC via the function closures.
+local function copy_set(S)
+    local new_set = {}
+    for v,_ in pairs(S) do
+        new_set[v] = true
+    end
+    return new_set
+end
 
-    local defined_variables = {} -- { var_id }, sorted by first definition
-    local last_use          = {} -- { var_id => integer }
-    local first_definition  = {} -- { var_id => integer }
+local function flow_analysis(block_list, state_list)
+    local function apply_gen_kill_sets(flow_state)
+        local input = flow_state.input
+        local output = flow_state.output
+        local gen = flow_state.gen
+        local kill = flow_state.kill
+        local in_changed = false
 
-    for i, cmd in ipairs(flat_cmds) do
-        for _, val in ipairs(ir.get_srcs(cmd)) do
-            if val._tag == "ir.Value.LocalVar" then
-                local v_id = val.id
-                last_use[v_id] = i
+        for v, _ in pairs(output) do
+            local val = true
+            if kill[v] then
+                val = nil
             end
+            local previous_val = input[v]
+            local new_val = previous_val or val
+            input[v] = new_val
+            in_changed = in_changed or (previous_val ~= new_val)
         end
-        for _, v_id in ipairs(ir.get_dsts(cmd)) do
-            local typ = func.vars[v_id].typ
-            if types.is_gc(typ) and not first_definition[v_id] then
-                first_definition[v_id] = i
-                table.insert(defined_variables, v_id)
+
+        for v, g in pairs(gen) do
+            assert(g ~= true or g ~= kill[v], "gen and kill can't both be true")
+            local previous_val = input[v]
+            local new_val = true
+            input[v] = new_val
+            in_changed = in_changed or (previous_val ~= new_val)
+        end
+
+        for v, _ in pairs(input) do
+            if not output[v] and not gen[v] then
+                input[v] = nil
+                in_changed = true
             end
         end
+
+        return in_changed
     end
 
-    -- 2) Find which variables are live at each GC spot in the program.
-
-    local live_gc_vars = {} -- { cmd => {var_id}? }
-    for i, cmd in ipairs(flat_cmds) do
-        local tag = cmd._tag
-        if
-            tag == "ir.Cmd.CallStatic" or
-            tag == "ir.Cmd.CallDyn" or
-            tag == "ir.Cmd.CheckGC"
-        then
-            live_gc_vars[cmd] = {}
-            for _, v_id in ipairs(defined_variables) do
-                local a = first_definition[v_id]
-                local b = last_use[v_id]
-                if a and b and a < i and i <= b then
-                    table.insert(live_gc_vars[cmd], v_id)
-                end
+    local function merge_live(input, output)
+        for v, _ in pairs(input) do
+            output[v] = true
+        end
+    end
+
+    local function empty_set(S)
```

Comment:
The name empty_set makes me think of function that returns a new empty_set. Perhaps rename to clear_set?

Reply:
Ok. Will call it clear_set.